### PR TITLE
Pass to lower linalg.fill and linalg.matmul to ukernel Op

### PR DESCRIPTION
**Pull request by grypp**
This PR introduces a pass that matches `linalg.matmul` and lowers into `ukernel.generic` Op.

Typically `ukernel.generic` needs shared memory space, so the pass allocates it by checking the tile size and pipeline stages.

If `linalg.matmul` has a `linalg.fill` producer  that is removed and fill value is passed to `ukernel.generic`. So it fuses.

If `linalg.matmul` has any consumer `ukernel.generic` has different name.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Passes.h [hunk 1]
```diff
@@ -552,6 +552,14 @@ std::unique_ptr<OperationPass<ModuleOp>> createConvertToNVVMPass();
 /// Performs the final conversion to ROCDL+LLVM dialect.
 std::unique_ptr<OperationPass<ModuleOp>> createConvertToROCDLPass();
 
+/// A Pass to lowers linalg's fill and matmul to microkernel op
+std::unique_ptr<OperationPass<func::FuncOp>> createLLVMGPULowerToUKernelsPass();
+
+/// Lowers linalg's fill and matmul to microkernel op
+FailureOr<Operation *> lowerMatmulToMicrokernel(
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Passes.h:559]: does this need to be exposed this way? This is GPU specific transformation.... There is a similar matcher for CPU, so it can be confusing.

I need to split this `Passes.h` into 4 parts.....

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/lower_ukernel.mlir [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        %11 = affine.apply affine_map<()[s0] -> ((s0 floordiv 32) * 128)>()[%workgroup_id_x]
+        %12 = affine.apply affine_map<()[s0] -> (s0 * 256 - (s0 floordiv 32) * 8192)>()[%workgroup_id_x]
+        flow.dispatch.tensor.store %10, %2, offsets = [%11, %12], sizes = [%c128, %c256], strides = [1, 1] : tensor<?x?xf32> -> !flow.dispatch.tensor<readwrite:tensor<8192x8192xf32>>
+        return
+      }
+    }
+  }
+}
+
+// CHECK: linalg.matmul
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/lower_ukernel.mlir:228]: Nit: new line at end of file

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+struct LLVMGPULowerToUKernelsPass
+    : LLVMGPULowerToUKernelsBase<LLVMGPULowerToUKernelsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Codegen::IREECodegenDialect>();
+    registry.insert<mlir::bufferization::BufferizationDialect>();
+  }
+  void runOnOperation() override;
+};
+
+static void generateMicrokernelName(std::string &fnName,
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp:36]: I am not sure the call should be so tuned the tile size... The tile size should be passed as arguments, and after inlining this should all become a no-op.....

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp (hunk 3 above)
Comment (grypp) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp:36]: While I understand that tiles sizes are eventually inlined, I wanted to clarify that we compile microkernels separately, and inlining happens after we linked them in the kernel. Tile sizes are needed in the compile-time while compiling microkernels.

Moreover, tile sizes are a template parameter in Cutlass classes. I cannot pass a function argument to them.

File: compiler/src/iree/compiler/Codegen/Passes.h (hunk 1 above)
Comment (grypp) [reply compiler/src/iree/compiler/Codegen/Passes.h:559]: I have implemented transform dialect op for microkernel that calls `lowerMatmulToMicrokernel` for transformation (that will come as next PR :)) .

I agree that it is GPU specific, how about I change its name to `lowerGPUMatmulToMicrokernel`?

Comment (grypp) [reply compiler/src/iree/compiler/Codegen/Passes.h:559]: ```I need to split this Passes.h into 4 parts.....```
This makes sense to me. I am happy to do it if you are busy :)

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp:36]: Ok, I understand the issue here, but I am concerned about the blow up in the number of "microkernels" that will have to be used... So this doesnt seem to get to a reasonable end-state... The cost of maintaining and shipping so many variants just on tile sizes is non-trivial. I guess the decision isnt upto me... We should escalate this to someone who can be the decision maker on this.

File: compiler/src/iree/compiler/Codegen/Passes.h (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Codegen/Passes.h:559]: If you are willing to do it, I'll be very happy!

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp (hunk 3 above)
Comment (grypp) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp:36]: We identify the most commonly used tile sizes, approximately 5, and generating microkernels for those sizes are sufficient. I mean we don't need every tile size.

Also, generating microkernels with different tile sizes is trivial. I've implemented generator that generates microkernel by adding single line.

**Comment by MaheshRavishankar** [comment]
Ok, the pass itself looks fine to me... but I'd want to get some more eyes on this path and see if this is something we want to support.... This is opening a door that can easily get out of hand.

@allieculp for visibility. This is relevant to the macro kernel vs micro kernel discussion.

**Comment by grypp** [comment]
> Ok, the pass itself looks fine to me... but I'd want to get some more eyes on this path and see if this is something we want to support.... This is opening a door that can easily get out of hand.


I'm wondering if we could consider making this PR as pass-only? This would allow me to put my follow-up PR for the transform dialect op on top of this one. 

I believe that having a microkernel pass would be a valuable developer tool, regardless of the specific support we want to provide. 

In my opinion, it would help us achieve optimal performance quickly, as it is currently on par with cublas for A100. However, I understand that some people may be hesitant to use it. 


> @allieculp for visibility. This is relevant to the macro kernel vs micro kernel discussion.

Let's discuss this further in the matmul meeting.

**Comment by MaheshRavishankar** [comment]
> > Ok, the pass itself looks fine to me... but I'd want to get some more eyes on this path and see if this is something we want to support.... This is opening a door that can easily get out of hand.
> 
> I'm wondering if we could consider making this PR as pass-only? This would allow me to put my follow-up PR for the transform dialect op on top of this one.
> 
> I believe that having a microkernel pass would be a valuable developer tool, regardless of the specific support we want to provide.

So you are saying add the pass in this PR, and add the method to do the transformation in the PR that adds a transform dialect op? That makes sense, but it will all be contingent on the decision made in the matmul meeting. So you can wait for that to make progress here since I dont want to waste your time.

**Comment by grypp** [comment]
During our meeting, several questions arose that I'd like to address in this pull request:

### **multiple microkernels** vs **single microkernel**
Current implementation requires **multiple microkernels** due to the templated function names based on microkernel properties. This necessitates generating multiple ABIs (I've implemented a generator for that, it is just a line of code to have a new microkernel). I believe it would be advantageous to track fast microkernels and fallback to iree codegen when a microkernel is not implemented. Additionally, the register size is controllable.

An alternative idea is to use a **single microkernel** that covers all cases. This approach is commonly used for CPUs but may not be ideal for GPUs. While having a single microkernel might simplify things, it could result in a large switch statement, potentially causing register size issues and making it incompatible with the nvcc compiler.

### Inline assembly 
Another question that came up was regarding the use of **inline assembly** (as MLIR op or, asm in CUDA). This poses a challenge since a GPU microkernel typically consists of approximately 4000 lines of PTX code. Assuming we have at least 50 microkernels for different data types and tiles, using inline assembly would require modifying the PTX code by removing register allocation and parameter loading sections. Additionally, any CUDA parameters in the inline assembly would need the '%' character changed to '%%' in the PTX code.

**TL;DR** I see little distinction between many and single microkernels. Many microkernels offer greater control without fallbacks and are more compatible with cutlass and CUDA toolchain. Regarding inline assembly, while desirable, automating modifications adds complexity.

**Comment by benvanik** [comment]
Closing stale PR.
